pkg/controllers/trace_controller.go

```diff
@@ -247,6 +254,7 @@ func (r *TraceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	}
 
 	// Call gadget operation
+	cleanupTraceStatus(trace)
 	traceBeforeOperation := trace.DeepCopy()
 	patch := client.MergeFrom(traceBeforeOperation)
 	gadgetOperation.Operation(req.NamespacedName.String(), trace)
```

Comment:
TBH, by analysing the code I immediately thought that the operations that do not set the `State` but just the `Output` (e.g. `generate` on seccomp or `update` on network-policy) will not work properly after this PR. However, I tested the seccomp gadget and it worked correctly: after the `generate` operation, the `State` remained on `Started`.

I think that [`Patch()`](https://github.com/kinvolk/inspektor-gadget/blob/755727cdc6b635d35cd776909c8e367a641e63e9/pkg/controllers/trace_controller.go#L72) does all the work. It seems that it patches the actual resource with the differences between the `patch` (which you generated **before** the clean up) and the `trace` which is updated by the `gadgetOperation`. Therefore, any change on the `State` will trigger an update and, in the case the operation just changes the `Output`, the `State` will stay empty on `patch` and `trace` thus it will not trigger any update on the actual resource.

The only problem I see with this PR is that the logs could be misleading. They would look like the `State` is being cleaned up but actually, it means that the `State` will not be updated and will stay invariable on its current value:

```
time="2022-02-23T07:35:39Z" level=info msg="Gadget completed operation. Trace status will be updated accordingly"
time="2022-02-23T07:35:39Z" level=info msg="Updating new status of trace \"gadget/seccomp\": state= operationError=\"\" operationWarning=\"\" output=<353 characters>"
```

Reply:
> TBH, by analysing the code I immediately thought that the operations that do not set the `State` but just the `Output` (e.g. `generate` on seccomp or `update` on network-policy) will not work properly after this PR. However, I tested the seccomp gadget and it worked correctly: after the `generate` operation, the `State` remained on `Started`.
> 

I didn't consider this case. 

> I think that [`Patch()`](https://github.com/kinvolk/inspektor-gadget/blob/755727cdc6b635d35cd776909c8e367a641e63e9/pkg/controllers/trace_controller.go#L72) does all the work. It seems that it patches the actual resource with the differences between the `patch` (which you generated **before** the clean up) and the `trace` which is updated by the `gadgetOperation`. Therefore, any change on the `State` will trigger an update and, in the case the operation just changes the `Output`, the `State` will stay empty on `patch` and `trace` thus it will not trigger any update on the actual resource.
> 

You're almost correct. I think the problem is that I'm generating the patch **after** the cleanup, so if the gadget doesn't set a status field then it's not updated at all (it was empty before and after and hence the patch method thinks there's nothing to do). 

> The only problem I see with this PR is that the logs could be misleading. They would look like the `State` is being cleaned up but actually, it means that the `State` will not be updated and will stay invariable on its current value:
> 

I think this version is wrong, it wont' actually cleanup any field. I swapped lines 256 and 257, but now it doesn't work. The `stop` operation will cleanup `Output` created by `generate`. I'm not really sure how to implement it them. Perhaps we can move `cleanupTraceStatus` back to the gadgets but calling it only once per operation where it makes sense.

Reply:
> I think the problem is that I'm generating the patch **after** the cleanup
>
It was _after_ in my mind but I wrote _before_ hahaha 🤦🏽

> I'm not really sure how to implement it them.
>
Me neither, the way `generate` and `update` operations work does not allow us to have an easy solution here 🙁 .

> Perhaps we can move `cleanupTraceStatus` back to the gadgets but calling it only once per operation where it makes sense.
Yeah, I think that's the only way out right now. It is a pity, though.

Reply:
I think I found a way. We don't need to clean the whole status of the trace, just cleaning `OperationError` and `OperationWarning` before calling the operation should be good enough. Gadgets should take care of cleaning/setting the `Output` as needed.